### PR TITLE
feat(store): implement external store machinery

### PR DESCRIPTION
**Pull request by michaelfig**
Refs #1846

Consolidate `@agoric/weak-store` with `@agoric/store` and add `makeExternalStore` and friends to its exports.

I'd like to land this in the SDK, once tests pass and everybody is content.

### CODE REVIEW

**Comment by michaelfig** [comment]
@dtribble, @erights please take a look at my update of [packages/ERTP/src/issuer.js](https://github.com/Agoric/agoric-sdk/pull/1843/files#diff-908a78b44aa919e08b9a20e447667f9cR6) to use the External Store API.  It looks like the changes are minimal.

@FUDCo, @warner please have a look at [packages/store/test/test-external-store.js](https://github.com/Agoric/agoric-sdk/pull/1843/files#diff-d989e0772a61667bd43a64ce2128bf3fR54) for how I supply hooks to `makeHydrateExternalStoreMaker` in order to use a (leaky) in-memory backend implementation.  I think the SwingSet integration can be implemented in terms of that function.

**Comment by michaelfig** [comment]
> I love it!

:smile:  Do note that I separated out `makeDepositFacet` since that would use a different External Store.  So, the change gained a few lines.

File: packages/store/src/types.js
Comment (erights) [inline packages/store/src/types.js]:
```suggestion
 * @property {(key: K, value: V) => void} init - Initialize the key only if it 
   doesn't already exist
```
and elsewhere.

File: packages/store/src/external/memory.js
Comment (erights) [inline packages/store/src/external/memory.js:19]: This makes beautifully clear that the external store really is just a logical extension of the heap, rather than something that visibly has collection-like properties. Here, there is no collection of made instances at all. They're just stored in the language's heap of objects. End of story. Beautiful!

File: packages/store/src/external/hydrate.js
Comment (erights) [inline packages/store/src/external/hydrate.js]: If this is intended to demonstrate how one of these would work, fine. But it looks to me like this is intended to be production code. In which case, do *not* use `JSON.stringify`/`JSON.parse` directly. For one thing, it has too many irregularities even on primitive data types: It won't encode `NaN`, `Infinity`, `-Infinity`, `undefined`, or BigInts.

You can use marshal as a fancy data-only serializer by simply insisting that the marshal does not produce any slots, for example by providing a `valueToSlot` function that throws. We might extend from here to a useful valueToSlot in the future. But `JSON.serialize` doesn't handle those cases either, so a data-only marshal would still be superior in all ways to using raw `JSON.stringify`. 

@FUDCo , please double check what I said about JSON above. Thanks.

File: packages/store/src/external/hydrate.js
Comment (erights) [inline packages/store/src/external/hydrate.js]: I don't think this is a practical implementation of `markDirty`. We should serialize dirty data only once per transaction. OTOH this code itself is not in a position to determine what happens at end-of-transaction. But `makeBackingStore` is. Here, can we tell `store` simply that the `instanceKey` is dirty, combined with adding the `serialize` function as an argument to `backing.makeStore`, so the backing store can be told here what will need to be serialized, and how to serialize it, but leave it to the backing store to determine when to do so.

Which btw raises the SwingSet question of how the backing store will arrange to commit or abort atomically with the vat using it. I also have questions about what happens to backing store state during replay. I think the answer is nothing, since the backing store is a device. All answers from the backing store would be part of the vat's replay log, and the "real" backing store would not be reconnected until the vat replayed to that point. But that point does not include the writes from the aborted crank, so that will need to be absent from the committed backing store state. Thus, `makeBackingStore` must already be aware of transaction boundaries, and so can know when to serialize dirty state.

File: packages/store/src/external/hydrate.js
Comment (erights) [inline packages/store/src/external/hydrate.js]: The term "instanceName" confused me. In this scheme, when I hear "instance", I immediately think about the objects made by the maker, which are the arguments of `instanceToKey`. Can we call this something like `storeName`?

File: packages/store/src/external/hydrate.js
Comment (erights) [inline packages/store/src/external/hydrate.js]: Why `true` vs `undefined` rather than `true` vs `false`?

File: packages/store/src/external/hydrate.js
Comment (erights) [inline packages/store/src/external/hydrate.js]: Why do we prefer these keys to be strings rather than numbers?

File: packages/store/src/external/hydrate.js
Comment (erights) [inline packages/store/src/external/hydrate.js:102]: Nice!

**Review by erights — CHANGES_REQUESTED** [review]
This system is so beautiful, and solves such a crucial problem for us so simply, that I actually teared up reading this code. Really. I can't tell you how much I appreciate seeing this problem collapse into this solution. Thanks!

File: packages/store/src/external/hydrate.js
Comment (FUDCo) [reply packages/store/src/external/hydrate.js]: What @erights says about JSON is right.

Comment (michaelfig) [reply packages/store/src/external/hydrate.js]: Okay, I will add a dependency on `@agoric/marshal`.

File: packages/store/src/external/hydrate.js
Comment (michaelfig) [reply packages/store/src/external/hydrate.js]: > I don't think this is a practical implementation of `markDirty`. We should serialize dirty data only once per transaction.

Okay, I will add the necessary hooks to do that.

I think the serialization should only happen in the backing store, not this layer.  So I will push it down.

File: packages/store/src/external/hydrate.js
Comment (michaelfig) [reply packages/store/src/external/hydrate.js]: I use the `$hinit && expr` idiom, and I think it would be better to return `undefined` when the expression is short-circuited.  Later, I can imagine `$hinit` might be an object or function, but I want it to be `true` for now because I don't know what it would become yet.

File: packages/store/src/external/hydrate.js
Comment (michaelfig) [reply packages/store/src/external/hydrate.js]: > Why do we prefer these keys to be strings rather than numbers?

No particular reason.  I will adjust.

File: packages/store/src/external/hydrate.js
Comment (michaelfig) [reply packages/store/src/external/hydrate.js]: > This system is so beautiful, and solves such a crucial problem for us so simply, that I actually teared up reading this code. Really. I can't tell you how much I appreciate seeing this problem collapse into this solution. Thanks!

Thank you!  :blush: I'm really glad you insisted on exploring rewrites as an avenue.  The work is far from done, but we're getting there.

File: packages/store/src/external/hydrate.js
Comment (michaelfig) [reply packages/store/src/external/hydrate.js]: > Okay, I will add a dependency on `@agoric/marshal`.

Oh, with my note below, I will push the serialization down into the backing store rather than have it at this layer.  That will enable true flexibility.

File: packages/store/src/external/hydrate.js
Comment (erights) [reply packages/store/src/external/hydrate.js]: Good point about the short circuit. I agree.

File: packages/store/src/external/hydrate.js
Comment (erights) [reply packages/store/src/external/hydrate.js]: I like that plan.

File: packages/store/src/external/hydrate.js
Comment (michaelfig) [reply packages/store/src/external/hydrate.js]: > The term "instanceName" confused me. [...] Can we call this something like `storeName`?

I moved back to 'keyName', which is the same thing we used for the `makeWeakStore`/`makeStore` argument.  If you'd like to change that, we should change all three places.

File: packages/store/src/external/hydrate.js
Comment (michaelfig) [reply packages/store/src/external/hydrate.js]: I've improved the typing so that this argument is omitted for the rehydrate case rather than explicitly `undefined`.

File: packages/store/src/external/hydrate.js
Comment (erights) [reply packages/store/src/external/hydrate.js]: I still see `instanceName`. Did you forget to push? Unresolving.

**Review by erights — APPROVED** [review]
LGTM modulo stale name

File: packages/store/src/external/hydrate.js
Comment (michaelfig) [reply packages/store/src/external/hydrate.js]: Oops.  Fixed. :smile:

**Comment by FUDCo** [comment]
I'm presuming in all this there's some magic scheme I haven't yet learned about that will enable the implementation of a large weakStore backed by a database on disk?  Because off the top of my head I can't think of any way to possibly do that, but on the other hand I've already seen some magic here.

File: packages/store/src/weak-store.js
Comment (FUDCo) [inline packages/store/src/weak-store.js:22]: Was there consensus on the store API design before I joined?  Having `init` be distinct from `set` and having `delete` require the thing being deleted to first exist both seem like nods to purity that in my experience are ergonomic annoyances that people will just write wrappers around, resulting in a bunch of extra database accesses with no corresponding benefit.  Certainly the stores we use in the kernel don't have this.

File: packages/store/src/external/hydrate.js
Comment (FUDCo) [inline packages/store/src/external/hydrate.js:82]: In all the layering of makers of makers of makers of makers, I'm finding myself completely unable to discern where one injects the actual thing that stores the bits on disk.

File: packages/store/src/weak-store.js
Comment (erights) [reply packages/store/src/weak-store.js:22]: > Was there consensus on the store API design before I joined?

This was a feature of the store API from the beginning. It is the defining difference that caused us to coin the term "store" to distinguish it from JS's Maps. When did you join? That + the GitHub history can answer this question.

> my experience are ergonomic annoyances that people will just write wrappers around

For *defensive code* our experience is exactly the opposite. Tons of asserts that would have decorated each `get` and `set` magically disappeared by using stores rather than maps. I remember this being really striking in my early reviews of the Zoe implementation. Over and over again I would first think "we need to check here that...". Then I'd remember, "oh yeah, it is a store. It is already checking that."

File: packages/store/src/external/hydrate.js
Comment (erights) [reply packages/store/src/external/hydrate.js:82]: We should walk through this verbally together and consider alternative names to make the levels clearer. As you know from the E days, ocap style often more heavily curries making functions, because different steps towards instantiation have different authority issues. Better naming can indeed help a lot.

**Comment by michaelfig** [comment]
> I'm presuming in all this there's some magic scheme I haven't yet learned about that will enable the implementation of a large weakStore backed by a database on disk?

That's in the stub SwingSet implementation I did in #1857.  See [`packages/SwingSet/src/kernel/vatManager/externalStore.js`](https://github.com/Agoric/agoric-sdk/pull/1857/files#diff-8a1c73eb6d3f12247f6e67f50f9c8e66R11).  It's still in memory, but the rewrite and APIs are working.

Still have to extend the rewrite so that all of `issuer.js` at least passes without warning.

**Comment by FUDCo** [comment]
> > I'm presuming in all this there's some magic scheme I haven't yet learned about that will enable the implementation of a large weakStore backed by a database on disk?
> 
> That's in the stub SwingSet implementation I did in #1857. See [`packages/SwingSet/src/kernel/vatManager/externalStore.js`](https://github.com/Agoric/agoric-sdk/pull/1857/files#diff-8a1c73eb6d3f12247f6e67f50f9c8e66R11). It's still in memory, but the rewrite and APIs are working.

That appears to be a placeholder backed by a `WeakMap`.  I'm specifically asking about how we can have a weakStore that's _not_ in memory.  How we gonna do that?  I'm very nervous about this.

**Comment by erights** [comment]
The on-disk portion cannot be weak until we have in-ram WeakRefs. In memory WeakMaps reduce in-ram memory pressure. But they give us no gc info for us to make our own decisions about memory elsewhere. This is fundamental to the problem independent of choice of representation.

The large virtual weak store is a weak store in the sense that it provides the rights amplification properties that weakMaps provide that maps do not.

**Comment by FUDCo** [comment]
That's sort of what I thought, but I was hoping this would be part of that voodoo that you do.

If we don't worry about GC (heh) does a non-iterable strong map keyed by an unforgeable object identity provide the rights amplification mechanism we need?

**Comment by erights** [comment]
> If we don't worry about GC (heh) does a non-iterable strong map keyed by an unforgeable object identity provide the rights amplification mechanism we need?

yes. The on-disk portion of the virtual weakmap is equiv to that. But using a weakmap for the in-memory portion does relieve in-memory memory pressure while a rights-amplifying strong map would not.

**Comment by erights** [comment]
So many things will get so much better once we have deterministic WeakRefs.

File: packages/store/src/weak-store.js
Comment (dtribble) [reply packages/store/src/weak-store.js:22]: I wasn't involved the current `store` api, and did have some similar concerns. It was surprising how much code `init` in particular cleaned up for several use cases. 

The current API seems a little too limited to work for all our use cases, however. For example, we had some minor wallet bugs with the current semantics that had to be addressed by manually checking `has`. But it wasn't quite time to revisit the APIs until we had lots more examples, especially persistence. Now that we are doing persistence, we can revisit additional functionality. TBD whether it needs more operations, options in current operations, or a separate API for other use cases (e.g., a fancy query API). 

In both Xanadu and Midori collections, we had `get` vs `fetch` operations where `get` requires a value and `fetch` doesn't. My favorite missing operation for normal collection usage is an an efficient `provide` operation that returns a value if it's present or runs some code to produce it if it's not. In Smalltalk, that's the base operation that provides flexibility and reliability. In our `store`, that could be implemented with an optional second `ifAbsent` argument that is invoked if the object is not there.

Comment (michaelfig) [reply packages/store/src/weak-store.js:22]: > or a separate API for other use cases (e.g., a fancy query API).

I hope to have other data structures exposed by the external store, such as a `makePriorityQueue(comparator)`, or somesuch.  I don't want to add more functionality to the `makeWeakStore()` API itself, unless it can be easily done for the in-memory version of the WeakStore, too.

> In our store, that could be implemented with an optional second ifAbsent argument that is invoked if the object is not there.

That sounds good!

File: packages/store/src/external/hydrate.js
Comment (warner) [inline packages/store/src/external/hydrate.js:66]: I'm going to need a walkthrough to be confident that I understand this code, but this fetch has me concerned. From our other discussions about the "representative", we know there will be multiple representatives for the same virtual object, because the code we hand one to might hold on to it for an arbitrariy long period of time (and we have no way to tell if/when they've stopped using it). If the data is only fetched from the secondary store once, when we create the representative, it will be stale if/when some other user modifies their own representative.

My hunch is that we must be maximally lazy on read, and have every getter invocation cause a read from secondary storage, to make sure that every `get()` sees the latest value.

File: packages/store/src/external/hydrate.js
Comment (warner) [inline packages/store/src/external/hydrate.js:62]: Am I correct in thinking that liveslots will somehow get access to this `load` method, and will convert inbound references to our "virtual objects" into calls to `load()`?

**Review by warner — COMMENTED** [review]
Sounds promising to me, but I'll need to be walked through the likely liveslots interface.

File: packages/store/src/external/hydrate.js
Comment (michaelfig) [reply packages/store/src/external/hydrate.js:66]: Yes, you're absolutely right.  I'll make the "active record" something that SwingSet can implement as desired.  That will also solve my problem with the current test (divergence between representatives).

File: packages/store/src/external/hydrate.js
Comment (michaelfig) [reply packages/store/src/external/hydrate.js:62]: Correct.  Please see the next PR in this series, #1857.

**Pull request merged**
